src/cpp/src/multinomial_decoding.cpp

```diff
@@ -0,0 +1,271 @@
+
+    float prefix_sum = 0.0f;
+
+    // top_p
+    for (int i = 0; i < vocab_size; i++) {
+        prefix_sum += token_scores[i].score;
+        if (prefix_sum >= top_p) {
+            return first + (i + 1);
+        }
+    }
+
+    return last;
+}
+
+void sampling_repetition_penalty(float* first, float* last, const std::vector<int64_t>& input_ids, float penalty) {
+    const float inv_penalty = 1.f / penalty;
+    const int vocab_size = last - first;
+    std::vector<bool> occurrence(vocab_size, false);
+    for (const int64_t id : input_ids) {
+        if (!occurrence[id]) {
+            first[id] *= (first[id] > 0) ? inv_penalty : penalty;
+        }
+        occurrence[id] = true;
+    }
+}
+
+void sampling_temperature(float* first, float* last, float temp) {
+    const float inv_temp = 1.f / temp;
+    for (float* it = first; it != last; it++) {
+        *it *= inv_temp;
+    }
+}
+
+struct SamplingParameters {
+    int top_k;
+    float top_p;
+    float temperature;
+    float repetition_penalty;
+
+    SamplingParameters(ov::GenerationConfig generation_config) {
+        // parameters validation
+        OPENVINO_ASSERT(generation_config.top_k > 0,
+                        "top_k must be a strictly positive float, but got ",
+                        generation_config.top_p);
+        OPENVINO_ASSERT(generation_config.top_p > 0 || generation_config.top_p < 1.0f,
+                        "top_p must be a positive float > 0 and < 1, but got ",
+                        generation_config.top_p);
+        OPENVINO_ASSERT(generation_config.temperature > 0,
+                        "Temperature must be a strictly positive float, but got ",
+                        generation_config.temperature);
+        OPENVINO_ASSERT(generation_config.repetition_penalty > 0,
+                        "Repetition penalty must be a strictly positive float, but got ",
+                        generation_config.repetition_penalty);
+
+        top_k = generation_config.top_k;
+        top_p = generation_config.top_p;
+        temperature = generation_config.temperature;
+        repetition_penalty = generation_config.repetition_penalty;
+    }
+};
+
+struct RandomSampling {
+    SamplingParameters parameters;
+    RandomSampling(SamplingParameters parameters) : parameters{std::move(parameters)} {}
+
+    TokenIdScore get_out_token(float* logits, size_t vocab_size, std::vector<int64_t> tokens) {
+        // logits pre-process
+        if (parameters.repetition_penalty != 1.0f) {
+            sampling_repetition_penalty(logits, logits + vocab_size, tokens, parameters.repetition_penalty);
+        }
+
+        if (parameters.temperature > 0) {
+            sampling_temperature(logits, logits + vocab_size, parameters.temperature);
+        }
+
+        std::vector<TokenIdScore> token_scores(vocab_size);
+        for (int i = 0; i < vocab_size; i++) {
+            token_scores[i] = TokenIdScore(i, logits[i]);

```

Comment:
```suggestion
            token_scores[i] = TokenIdScore{i, logits[i]};
```
Relates to constructor deletion.

Reply:
applied